File: local/src/com/hendraanggrian/local/internal/LocalBinding.java
package com.hendraanggrian.local.internal;

import com.hendraanggrian.local.LocalSettings;
import com.hendraanggrian.local.Saver;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@SuppressWarnings("unused")
public abstract class LocalBinding implements Saver {

    @SuppressWarnings("WeakerAccess") protected final LocalSettings<?> source;

    protected LocalBinding(@NotNull LocalSettings<?> source) {
        this.source = source;
    }

    @Nullable
    protected String getValue(@NotNull String key, @Nullable String defaultValue) {
        return source.getString(key, defaultValue);
    }

    protected int getValue(@NotNull String key, int defaultValue) {
        return source.getInt(key, defaultValue);
    }

    protected LocalSettings.Editor getEditor() {
        return source.getEditor();
    }

    protected void setValue(
        @NotNull LocalSettings.Editor editor,
        @NotNull String key,
        @Nullable String value
    ) {
        editor.setString(key, value);
    }

    protected void setValue(
        @NotNull LocalSettings.Editor editor,
        @NotNull String key,
        int value
    ) {
        editor.setInt(key, value);
    }
}

File: local/src/com/hendraanggrian/local/internal/LocalBinding.kt
package com.hendraanggrian.local.internal

import com.hendraanggrian.local.Local

abstract class LocalBinding(protected val source: Local<*>) :
    Local.Saver {

    protected fun get(key: String, defaultValue: String?): String? =
        source.getString(key, defaultValue)

    protected fun get(key: String, defaultValue: Int): Int =
        source.getInt(key, defaultValue)

    protected fun get(key: String, defaultValue: Long): Long =
        source.getLong(key, defaultValue)

    protected fun get(key: String, defaultValue: Float): Float =
        source.getFloat(key, defaultValue)

    protected fun get(key: String, defaultValue: Boolean): Boolean =
        source.getBoolean(key, defaultValue)

    protected fun getEditor(): Local.Editor = source.getEditor()

    protected fun set(editor: Local.Editor, key: String, value: String?) {
        editor[key] = value
    }

    protected fun set(editor: Local.Editor, key: String, value: Int) {
        editor[key] = value
    }

    protected fun set(editor: Local.Editor, key: String, value: Long) {
        editor[key] = value
    }

    protected fun set(editor: Local.Editor, key: String, value: Float) {
        editor[key] = value
    }

    protected fun set(editor: Local.Editor, key: String, value: Boolean) {
        editor[key] = value
    }
}